firebase-segmentation/src/main/java/com/google/firebase/segmentation/CustomInstallationIdCache.java

```diff
@@ -0,0 +1,109 @@
+
+package com.google.firebase.segmentation;
+
+import android.content.SharedPreferences;
+import androidx.annotation.Nullable;
+import androidx.annotation.VisibleForTesting;
+import com.google.android.gms.common.util.Strings;
+import com.google.firebase.FirebaseApp;
+
+class CustomInstallationIdCache {
+
+  // Status of each cache entry
+  // NOTE: never change the ordinal of the enum values because the enum values are stored in cache
+  // as their ordinal numbers.
+  enum CacheStatus {
+    // Cache entry is synced to Firebase backend
+    SYNCED,
+    // Cache entry is waiting for Firebase backend response or pending internal retry for retryable
+    // errors.
+    PENDING,
+    // Cache entry is not accepted by Firebase backend.
+    ERROR,
+  }
+
+  private static final String SHARED_PREFS_NAME = "CustomInstallationIdCache";
+
+  private static final String CUSTOM_INSTALLATION_ID_KEY = "Cid";
+  private static final String INSTANCE_ID_KEY = "Iid";
+  private static final String CACHE_STATUS_KEY = "Status";
+
+  private static CustomInstallationIdCache singleton = null;
+  private final SharedPreferences prefs;
+
+  static CustomInstallationIdCache getInstance() {
+    if (singleton == null) {
+      singleton = new CustomInstallationIdCache();
+    }
+    return singleton;
+  }
+
+  private CustomInstallationIdCache() {
+    // Since different FirebaseApp in the same Android application should have the same application
+    // context and same dir path, so that use the context of the default FirebaseApp to create the
+    // shared preferences.
+    prefs =
+        FirebaseApp.getInstance()
+            .getApplicationContext()
+            .getSharedPreferences(SHARED_PREFS_NAME, 0); // private mode
+  }
+
+  @Nullable
+  synchronized CustomInstallationIdCacheEntryValue readCacheEntryValue(FirebaseApp firebaseApp) {
+    String cid =
+        prefs.getString(getSharedPreferencesKey(firebaseApp, CUSTOM_INSTALLATION_ID_KEY), null);
+    String iid = prefs.getString(getSharedPreferencesKey(firebaseApp, INSTANCE_ID_KEY), null);
+    int status = prefs.getInt(getSharedPreferencesKey(firebaseApp, CACHE_STATUS_KEY), -1);
+
+    if (Strings.isEmptyOrWhitespace(cid) || Strings.isEmptyOrWhitespace(iid) || status == -1) {
+      return null;
+    }
+
+    return CustomInstallationIdCacheEntryValue.create(cid, iid, CacheStatus.values()[status]);
+  }
+
+  synchronized void insertOrUpdateCacheEntry(
+      FirebaseApp firebaseApp, CustomInstallationIdCacheEntryValue entryValue) {
+    SharedPreferences.Editor editor = prefs.edit();
+    editor.putString(
+        getSharedPreferencesKey(firebaseApp, CUSTOM_INSTALLATION_ID_KEY),
+        entryValue.getCustomInstallationId());
+    editor.putString(
+        getSharedPreferencesKey(firebaseApp, INSTANCE_ID_KEY), entryValue.getFirebaseInstanceId());
+    editor.putInt(
+        getSharedPreferencesKey(firebaseApp, CACHE_STATUS_KEY),
+        entryValue.getCacheStatus().ordinal());
+    editor.commit();
```

Comment:
Will this code be called from the UI thread? If so, this will do blocking disk IO, which is generally discouraged. A few options to deal with this:
1. change `commit() -> apply()`, this will apply the change to disk asynchronously(no way to know if it failed)
2. make this method async e.g. returning `Task<Boolean>` and have a backing thread poll for IO operations

wdyt?